flask_resources/parsers/base.py

```diff
@@ -1,115 +1,157 @@
+The parser is not meant to parse the request body. For that you should use the
+``RequestBodyParser``.
 
+The request parser can accept both a schema or a dictionary. Using the schema
+enables you to do further pre/post-processing of values, while the dict version
+can be more compact.
 
-def select_request_parser(
-    resource_method, config_parser_or_parsers, default_parser_cls
-):
-    """Returns request parser corresponding to situation."""
-    if isinstance(config_parser_or_parsers, default_parser_cls):
-        return config_parser_or_parsers
+Example with schema:
 
-    return config_parser_or_parsers.get(
-        resource_method,
-        default_parser_cls(allow_unknown=False),  # default "parse nothing" parser
-    )
+.. code-block:: python
 
+    class MyHeaders(ma.Schema):
+        content_type = ma.fields.String()
 
-def request_parser_decorator(parser_cls, config_attr, context_attr):
-    """Decorator that parses the request object for the view."""
+    parser = RequestParser(MyHeaders, location='headers')
+    parser.parse()
 
-    def decorator(f):
-        @wraps(f)
-        def inner(self, *args, **kwargs):
-            parser = select_request_parser(
-                self.resource_method,
-                getattr(self.resource.config, config_attr),
-                parser_cls,
-            )
-            setattr(resource_requestctx, context_attr, parser.parse())
-            return f(self, *args, **kwargs)
+Same example with dict:
 
-        return inner
+.. code-block:: python
 
-    return decorator
+    parser = RequestParser({
+        'content_type': ma.fields.String()
+    }, location='headers')
+    parser.parse()
 
+**URL args parsing**
 
-class _ExcludeSchema(ma.Schema):
-    class Meta:
-        unknown = ma.EXCLUDE
+If you are parsing URL args, be aware that a query string can have repeated
+variables (e.g. in ``?type=a&type=b`` the value ``type`` is repeated).
 
+Thus if you build your own schema for URL args, you should inherit from
+``MultiDictSchema``. If you don't have repeated keys you can use a normal
+Marshmallow schema.
 
-class RequestParser:
-    """Request parser.
+**Unknown values**
+
+If you pass a dict for the schema, you can control what to do with unknown
+values:
+
+.. code-block:: python
 
-    Base class for parsing request.
-    """
+    parser = RequestParser({
+        'id': ma.fields.String()
+    }, location='args', unknown=ma.RAISE)
+    parser.parse()
 
-    DEFAULT_SCHEMA_CLASS = _ExcludeSchema
-    """The schema class to use if a dictionary is passed."""
+If you build your own schema, the same can be achieved with by providing the
+meta class:
 
-    def __init__(self, schema=None, allow_unknown=None):
+.. code-block:: python
+
+    class MyArgs(ma.Schema):
+        id = ma.fields.String()
+
+        class Meta:
+            unknown = ma.INCLUDE
+"""
+
+import marshmallow as ma
+from flask import request
+
+from .schema import MultiDictSchema
+
+
+class RequestParser:
```

Comment:
Replacing all parsers with a single one. Looks a bit in the direction of ``webargs``, but is simpler and will likely cause us less issues because we had problems with webargs in the past.